### ISSUE TITLE
Import issues

### ISSUE THREAD

**Issue by vdbergh** [glinscott/fishtest#521]
I think that a package in python generally should be self contained (I may be wrong, I have not checked other projects). In any case this used to be the case with the stats package.

But now for example if I do

``` fishtest@pt:~/fishtest/fishtest/fishtest/stats$ python stat_util.py ```

I get

```Traceback (most recent call last):
  File "stat_util.py", line 5, in <module>
    from fishtest.stats import LLRcalc
  File "/home/fishtest/fishtest/fishtest/fishtest/__init__.py", line 8, in <module>
    from fishtest.rundb import RunDb
  File "/home/fishtest/fishtest/fishtest/fishtest/rundb.py", line 17, in <module>
    from fishtest.views import format_results
  File "/home/fishtest/fishtest/fishtest/fishtest/views.py", line 12, in <module>
    import html
ImportError: No module named html
```

This is quite baffling to me since all the dependencies for properly executing the command are in the stats package itself. Is this a bug? Otherwise what is the rationale for the new behavior?

**Comment by tomtor** [comment]
@vdberg You need python >= 3.2 for the html package. Nothing has changed otherwise, except that views.py has this new dependency.

__init__.py is read, because it is in your PYTHONPATH 
or `from fishtest.stats ...` would fail

**Comment by vdbergh** [comment]
But I do not understand why the fishtest.stats... imports are necessary inside the stats package.  This wasn't like this before.

**Comment by vdbergh** [comment]
BTW I tried to run pserve under python3 but it did not work. I got the error

```Traceback (most recent call last):
  File "/home/fishtest/.local/bin/pserve", line 11, in <module>
    sys.exit(main())
  File "/home/fishtest/.local/lib/python3.5/site-packages/pyramid/scripts/pserve.py", line 34, in main
    return command.run()
  File "/home/fishtest/.local/lib/python3.5/site-packages/pyramid/scripts/pserve.py", line 275, in run
    app = loader.get_wsgi_app(app_name, config_vars)
  File "/home/fishtest/.local/lib/python3.5/site-packages/plaster_pastedeploy/__init__.py", line 129, in get_wsgi_app
    global_conf=defaults,
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 253, in loadapp
    return loadobj(APP, uri, name=name, **kw)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 277, in loadobj
    global_conf=global_conf)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 302, in loadcontext
    global_conf=global_conf)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 326, in _loadconfig
    return loader.get_context(object_type, name, global_conf)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 459, in get_context
    section)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 481, in _context_from_use
    object_type, name=use, global_conf=global_conf)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 412, in get_context
    global_conf=global_conf)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 302, in loadcontext
    global_conf=global_conf)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 334, in _loadegg
    return loader.get_context(object_type, name, global_conf)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 625, in get_context
    object_type, name=name)
  File "/home/fishtest/.local/lib/python3.5/site-packages/paste/deploy/loadwsgi.py", line 645, in find_egg_entry_point
    pkg_resources.require(self.spec)
  File "/home/fishtest/.local/lib/python3.5/site-packages/pkg_resources/__init__.py", line 901, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/home/fishtest/.local/lib/python3.5/site-packages/pkg_resources/__init__.py", line 787, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'fishtest_server' distribution was not found and is required by the application
```

**Comment by tomtor** [comment]
> But I do not understand why the fishtest.stats... imports are necessary inside the stats package. This wasn't like this before.

That is needed for python3 when these are imported from the `..` directory.

For pserve you need:

    python setup.py install
    pserve development.ini

Note that executing `python3 stat_util.py` also runs just fine after executing the setup

**Comment by vdbergh** [comment]
@tomtor  That is needed for python3 when these are imported from the .. directory.

Are you absolutely sure this cannot be solved in a more pythonic way? stats is independent of fishtest. I find it very strange that it would have to import things from itself via fishtest.stats...  As I told you in python2 the direct import worked fine.

python3 is working now. Thanks.

**Comment by vdbergh** [comment]
I will think about it. But next week I will be busy.

**Comment by vdbergh** [comment]
I quickly changed all the imports to direct imports (without fishtest.stats) and restarted pserve. Everything seems to be working normally. I did not start any tests though (it is bit too late for that).

python3 stat_util.py executes without any delay now.

**Comment by vdbergh** [comment]
Of course I only removed the fishtest.stats imports inside the stats package. Not elsewhere!

**Comment by tomtor** [comment]
See also https://stackoverflow.com/questions/12172791/changes-in-import-statement-python3

If tests fail we could prefix the import with `.`

**Comment by vdbergh** [comment]
Thanks! I like using "." I will change it in this way. Although I guess this would break python2 compatibility for stats :(

Still I don't understand then why everything is working normally with python2 style direct imports.... This morning I started a test and and everything is fine too.

**Comment by vdbergh** [comment]
It seems "from .blah import Blah" should also work in python2. I will change it. I was completely unaware of this syntax. Thanks for clarifying!

But I think they really solved a non-problem here. If I have module "string" in my package then of course I want that one and not the standard one.

**Comment by tomtor** [comment]
I am sure that just removing `fishtest` breaks pserve. I added it because of actual runtime errors.

**Comment by tomtor** [comment]
@vdbergh I just verified and the `.` is really needed. Note that after editting you must run

    python setup.py install

again, or pserve will use the old uneditted Python files

**Comment by vdbergh** [comment]
You are right. 

``` python3 setup.py install ```

appears to have installed fishtest globally

```
/usr/local/lib/python3.5/dist-packages/fishtest_server-0.1-py3.5.egg/fishtest/stats
```

Of course it did not react to changes in my local copy in ``` /home/fishtest```. 

I guess I need to use something like "virtualenv".

**Comment by tomtor** [comment]
> I guess I need to use something like "virtualenv".

Even then you need to run `setup.py install`

**Comment by vdbergh** [comment]
Yes now I was able to crash fishtest :)

I assume that ```from . import bla``` and friends will work, but it does not achieve the goal of being able to run ```stat_util.py``` without importing the full Fishtest framework :(
```
fishtest@pt:~/fishtest/fishtest/fishtest/stats$ python3 stat_util.py
Traceback (most recent call last):
  File "stat_util.py", line 9, in <module>
    from . import LLRcalc
SystemError: Parent module '' not loaded, cannot perform relative import
```
I guess they really really want you to use absolute imports.

Anyway ```import stats``` in the parent directory will work without importing Fishtest.

**Comment by vdbergh** [comment]
@tomtor For my python2 install (which no longer works of course) I have to run ```python setup.py install``` only once (afterwards it uses the local copy of fishtest). 

For my python3 install I need to run it every time (and it uses a system wide copy of fishtest) which is very inconvenient. Any idea what might cause this?

The only thing I do is changing ```#!/usr/bin/python2``` in  ```/home/fishtest/.local/bin/pserve``` (which is what ```pserve``` points to).

**Comment by vdbergh** [comment]
Anyway relative imports via ``` from . import bla``` and friends work. In the end I think I prefer this since it is more robust (it makes the stats package relocatable). I deleted all the ```if __name__==`__main__` ```: since this no longer works. I will make a PR soon.

**Comment by ppigazzini** [comment]
> I guess I need to use something like "virtualenv".

I suggest to use/read the script from the wiki:
https://github.com/glinscott/fishtest/wiki/Fishtest-server-setup#fishtest-server-script-to-test-a-prbranch

**Comment by ppigazzini** [comment]
Google Python Style Guide indicate that the short form of the import should be used only for third party or top level packages

http://google.github.io/styleguide/pyguide.html#233-decision

**Comment by vdbergh** [comment]
These are the key points (from "CONS")

> **Makes it harder to deploy code** because you have to replicate the package hierarchy. Not really a problem with **modern deployment mechanisms**.

I think the Google guidelines cater more to huge enterprise style monolithic software frameworks.

I had a little bit of experience with go packages and I found them very hard to use precisely because of the straight jacket they put you in.

**Comment by vdbergh** [comment]
Suppose that in a hierarchical file system we would suddenly be forced to use only absolute paths... 

Not a problem I guess with a "modern deployment mechanism".

**Comment by vdbergh** [comment]
I would not like relative import from a different package (like ```from import ..blab.bla import blah```) . However from the current package (```from . import blah```) it seems very natural...

**Comment by ppigazzini** [comment]
Import according to PEP8
https://pep8.org/#imports

> However, explicit relative imports are an acceptable alternative to absolute imports, especially when dealing with complex package layouts where using absolute imports would be unnecessarily verbose:
> ```python
> from . import sibling
> from .sibling import example
> ```

**Comment by vdbergh** [comment]
@ppigazzini As it seems to be controversial I closed the PR.

**Comment by ppigazzini** [comment]
@vdbergh IMO this issue is very interesting. Today, after gaining a better knowledge of venv/import. I simplified my helper scripts from:
```python
#!/usr/bin/python
import os, sys
sys.path.append(os.path.expanduser('~/fishtest/fishtest/fishtest'))
from rundb import RunDb
```
to:
```python
#!/usr/bin/env python
from fishtest.rundb import RunDb
```

**Comment by ppigazzini** [comment]
@vdbergh in this moment I don't have a good-enough knowledge of the topic to decide if/when use relative imports in the codebase, opinions are welcome :)

**Comment by tomtor** [comment]
@ppigazzini @vdbergh Just learned:

```python setup.py develop```

Allow in place editing without needing to run `setup.py` again before you restart `pserve`

**Comment by ppigazzini** [comment]
Tu quoque, @tomtor, fili mi! :)

```bash
${VENV}/bin/pip install --upgrade pip setuptools wheel
cd ${HOME}/fishtest/fishtest
${VENV}/bin/pip install -e .
```

https://stackoverflow.com/questions/45426794/setup-in-virtualenv-pip-install-e-vs-python-setup-py-install/45427228

**Comment by ppigazzini** [comment]
@tomtor @vdbergh I learned the hard way to always use python with virtual environment, now easier with python3. The administration of the old server was a nightmare, with packages installed either with the distro package manager and with pip at system level. And I fear to have done major damages with the first version of the wiki setup script where I used that horrible practice.
I read somewhere that a kitty die for any python package installed at system level...

By the way at first I found baroque the syntax used  for virtual environment in Pyramid documentation, but I confirm that it's very useful for communication with other people (scripts, wiki, issues, PRs etc.) 

> Why use `$VENV/bin/pip` instead of `source bin/activate`, then `pip`?
> `$VENV/bin/pip` clearly specifies that `pip` is run from within the virtual environment and not at the system level.

https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/install.html
